File: iosApp/iosApp/Modules/MnemonicUpdate ✅/MnemonicUpdateWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3lib

protocol MnemonicUpdateWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: MnemonicUpdateContext
    ) -> MnemonicUpdateWireframe
}

// MARK: - DefaultMnemonicWireframeFactory

final class DefaultMnemonicUpdateWireframeFactory {
    private let keyStoreService: KeyStoreService
    private let authenticateWireframeFactory: AuthenticateWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    
    init(
        keyStoreService: KeyStoreService,
        authenticateWireframeFactory: AuthenticateWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory
    ) {
        self.keyStoreService = keyStoreService
        self.authenticateWireframeFactory = authenticateWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
    }
}

// MARK: - MnemonicWireframeFactory

extension DefaultMnemonicUpdateWireframeFactory: MnemonicUpdateWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: MnemonicUpdateContext
    ) -> MnemonicUpdateWireframe {
        DefaultMnemonicUpdateWireframe(
            parent,
            context: context,
            keyStoreService: keyStoreService,
            authenticateWireframeFactory: authenticateWireframeFactory,
            alertWireframeFactory: alertWireframeFactory
        )
    }
}

// MARK: - Assembler

final class MnemonicUpdateWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> MnemonicUpdateWireframeFactory in
            DefaultMnemonicUpdateWireframeFactory(
                keyStoreService: resolver.resolve(),
                authenticateWireframeFactory: resolver.resolve(),
                alertWireframeFactory: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/mnemonicupdate/MnemonicUpdateWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicupdate

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.alert.AlertWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.mnemonicUpdate.MnemonicUpdateWireframe
import com.sonsofcrypto.web3walletcore.modules.mnemonicUpdate.MnemonicUpdateWireframeContext
import smartadapter.internal.extension.name

interface MnemonicUpdateWireframeFactory {
    fun make(parent: Fragment?, context: MnemonicUpdateWireframeContext): MnemonicUpdateWireframe
}

class DefaultMnemonicUpdateWireframeFactory(
    private val alertWireframeFactory: AlertWireframeFactory,
    private val keyStoreService: KeyStoreService,
): MnemonicUpdateWireframeFactory {

    override fun make(
        parent: Fragment?, context: MnemonicUpdateWireframeContext
    ): MnemonicUpdateWireframe = DefaultMnemonicUpdateWireframe(
        parent?.let { WeakRef(it) },
        context,
        alertWireframeFactory,
        keyStoreService,
    )
}

class MnemonicUpdateWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(MnemonicUpdateWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultMnemonicUpdateWireframeFactory(
                it.resolve(AlertWireframeFactory::class.name),
                it.resolve(KeyStoreService::class.name),
            )
        }
    }
}